test/ShallowWrapper-spec.js

```diff
@@ -2455,6 +2455,575 @@ describe('shallow', () => {
+          }
+          componentWillUpdate(nextProps, nextState, nextContext) {
+            spy('componentWillUpdate', this.props, nextProps, this.state, nextState, nextContext);
+          }
+          componentDidUpdate(prevProps, prevState, prevContext) {
+            spy('componentDidUpdate', prevProps, this.props, prevState, this.state, prevContext);
+          }
+          render() {
+            return <div>foo</div>;
+          }
+        }
+        Foo.contextTypes = {
+          foo: React.PropTypes.string,
+        };
+
+        const wrapper = shallow(
+          <Foo foo="bar" />,
+          {
+            context: { foo: 'context' },
+            lifecycleExperimental: true,
+          }
+        );
+        wrapper.setProps({ foo: 'baz' });
+        expect(spy.args).to.deep.equal(
+          [
+            [
+              'shouldComponentUpdate',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'state' }, { foo: 'state' },
+              { foo: 'context' },
+            ],
+            [
+              'componentWillReceiveProps',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'context' },
+            ],
+            [
+              'componentWillUpdate',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'state' }, { foo: 'state' },
+              { foo: 'context' },
+            ],
+            [
+              'componentDidUpdate',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'state' }, { foo: 'state' },
+              { foo: 'context' },
+            ],
+          ]
+        );
+      });
+
+      it('should cancel rendering when Component returns false in shouldComponentUpdate', () => {
+        const spy = sinon.spy();
+
+        class Foo extends React.Component {
+          shouldComponentUpdate() {
+            return false;
+          }
+          componentWillUpdate() {
+            spy();
+          }
+          componentDidUpdate() {
+            spy();
+          }
+          render() {
+            return <div>foo</div>;
+          }
+        }
+
+        const wrapper = shallow(<Foo foo="bar" />, { lifecycleExperimental: true });
+        wrapper.setProps({ foo: 'baz' });
+        expect(spy.callCount).to.equal(0);
+      });
+
+      it('should be batching updates in componentWillReceiveProps', () => {
+        const spy = sinon.spy();
+        class Foo extends React.Component {
+          constructor(props) {
+            super(props);
+            this.state = {
+              count: 0,
+            };
+          }
+          componentWillReceiveProps() {
+            this.setState({ count: this.state.count + 1 });
+            this.setState({ count: this.state.count + 1 });
+          }
+          render() {
+            spy();
+            return <div>{this.props.foo}</div>;
+          }
+        }
+        const result = shallow(<Foo />, { lifecycleExperimental: true });
+        result.setProps({ name: 'bar' });
```

Comment:
and `expect(spy).to.have.property('callCount', 1)` to have a nicer failure message ;-)